core/codegen/builder.go

```diff
@@ -0,0 +1,246 @@
+	}
+	endBlock := b.m.ctx.AddBasicBlock(b.function.llvm, "end_if")
+	if onFalse == nil {
+		falseBlock = endBlock
+	}
+
+	b.llvm.CreateCondBr(cond.llvm, trueBlock, falseBlock)
+
+	b.block(trueBlock, func() {
+		onTrue()
+		b.llvm.CreateBr(endBlock)
+	})
+
+	if onFalse != nil {
+		b.block(falseBlock, func() {
+			onFalse()
+			b.llvm.CreateBr(endBlock)
+		})
+	}
+
+	b.block(endBlock, func() {})
+}
+
+// While builds a logic block with the following psuedocode:
+//
+// while test() {
+//   loop()
+// }
+//
+func (b *Builder) While(test func() *Value, loop func()) {
+	testBlock := b.m.ctx.AddBasicBlock(b.function.llvm, "while_test")
+	loopBlock := b.m.ctx.AddBasicBlock(b.function.llvm, "while_loop")
+	exitBlock := b.m.ctx.AddBasicBlock(b.function.llvm, "while_exit")
+
+	b.llvm.CreateBr(testBlock)
+
+	b.block(testBlock, func() { b.llvm.CreateCondBr(test().llvm, loopBlock, exitBlock) })
+
+	b.block(loopBlock, func() {
+		loop()
+		b.llvm.CreateBr(testBlock)
+	})
+
+	b.block(exitBlock, nil)
+}
+
+// ForN builds a logic block with the following psuedocode:
+//
+// for it := 0; it < n; it++ {
+//   cont := cb()
+//   if cont == false { break; }
+// }
+//
+// If cb returns NoValue then the loop will never exit early.
+func (b *Builder) ForN(n *Value, cb func(iterator *Value) (cont *Value)) {
+	one := llvm.ConstInt(n.Type().llvmTy(), 1, false)
+	zero := b.Zero(n.Type())
+	iterator := b.LocalInit("loop_iterator", zero)
+
+	test := b.m.ctx.AddBasicBlock(b.function.llvm, "for_n_test")
+	loop := b.m.ctx.AddBasicBlock(b.function.llvm, "for_n_loop")
+	exit := b.m.ctx.AddBasicBlock(b.function.llvm, "for_n_exit")
+
+	b.llvm.CreateBr(test)
+
+	b.block(test, func() {
+		done := b.llvm.CreateICmp(llvm.IntSLT, iterator.Load().llvm, n.llvm, "for_n_condition")
+		b.llvm.CreateCondBr(done, loop, exit)
+	})
+
+	b.block(loop, func() {
+		it := iterator.Load()
+		cont := cb(it)
+		b.llvm.CreateStore(b.llvm.CreateAdd(it.llvm, one, "for_n_iterator_inc"), iterator.llvm)
+		if cont == nil {
+			b.llvm.CreateBr(test)
+		} else {
+			assertTypesEqual(cont.ty, b.m.Types.Bool)
+			b.llvm.CreateCondBr(cont.llvm, test, exit)
+		}
+	})
+
+	b.block(exit, nil)
+}
+
+type SwitchCase struct {
+	Conditions func() []*Value
+	Block      func()
+}
+
+// Switch builds a switch statement.
+func (b *Builder) Switch(val *Value, cases []SwitchCase, defaultCase func()) {
+	tests := make([]llvm.BasicBlock, len(cases))
+	blocks := make([]llvm.BasicBlock, len(cases))
+	for i := range cases {
+		tests[i] = b.m.ctx.AddBasicBlock(b.function.llvm, fmt.Sprintf("switch_case_%d_test", i))
+		blocks[i] = b.m.ctx.AddBasicBlock(b.function.llvm, fmt.Sprintf("switch_case_%d_block", i))
+	}
+
+	var defaultBlock llvm.BasicBlock
+	if defaultCase != nil {
+		defaultBlock = b.m.ctx.AddBasicBlock(b.function.llvm, "switch_case_default")
+		tests = append(tests, defaultBlock)
+	}
+
+	exit := b.m.ctx.AddBasicBlock(b.function.llvm, "end_switch")
+
+	b.llvm.CreateBr(tests[0])
+
+	for i, c := range cases {
+		i, c := i, c
```

Comment:
Looks weird to me, is this just creating a local copy, so that we don't modify the originals?

Reply:
Its the idiomatic way (unfortunately) to capture range iterators for lambdas and goroutines and the like. `i` and `c` here are more like iterator objects in that they get constructed once for the entirety of the loop, so if you use them in a deferred context (like the function below) you would only get the final value stored inside them. So this way instead you shadow the variable names with true local variables that get reference counted so you can actually use the values of the range within the function.

Reply:
It's idiomatic Go for loops containing closures that are executed after the for loop, as you'll get the iterator *pointer* captured, not the value. By creating another shadowed copy with the same name, the value is captured for each loop.

All that said, nothing in this block is executed after the for loop, so this line is actually pointless. Good spot.